go/engine/login_provision.go

```diff
@@ -917,6 +919,13 @@ func (e *loginProvision) ensurePaperKey(ctx *Context) error {
 		}
 	}
 
+	// Load me so that keys will be up to date.
+	var err error
+	e.arg.User, err = libkb.LoadUser(libkb.LoadUserArg{Self: true, UID: e.arg.User.GetUID(), PublicKeyOptional: true, Contextified: libkb.NewContextified(e.G())})
```

Comment:
Writing that dissertation made me realize it might not be. I'm going to try something. Meanwhile I'll leave it there so I can remember what's going on. You don't have to read it. :)

---

Our unsuspecting user first signs up on the website. Then they go to login on their first device with SharedDH enabled. They log in via username and password. `LoginProvision` has just provisioned this device using `Run->route->makeEldestDevice`. Now it's time for `ensurePaperKey` which decides to make a new one with `PaperKeyPrimary`. `PaperKeyGen` syncs the `SharedDHKeyring` (to be sure to encrypt all sdh key generations for the new paperkey). The `SharedDHKeyring` needs a `UPAK` to check that the sdh secretboxes came from the right folks. Where does it get this `UPAK`? "Normally" outside of login it would ask the `CachedUPAKLoader`, but the `CachedUPAKLoader` is indisposed*. So we need to pass a `UPAK` to the `SharedDHKeyring.Sync`. This `UPAK` comes from `PaperKeyGenArg.User.ExportToUPAK`. But if that `User` were loaded before `makeEldestDevice` then it wouldn't know about the device key encryption key, and the sync would fail. That is the reason that `User` is re-loaded here. So that the sync `UPAK` knows about the provisioned device key.

* If you use `CachedUPAKLoader` while holding a `LoginContext` it gets most of the way through and then soft-deadlocks on invalidating the `User` cache. I don't actually see a `LoginContext` here...